subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/execution/KotlinGrammar.kt

```diff
@@ -0,0 +1,282 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.kotlin.dsl.execution
+
+import org.jetbrains.kotlin.lexer.KtTokens.ANDAND
+import org.jetbrains.kotlin.lexer.KtTokens.ARROW
+import org.jetbrains.kotlin.lexer.KtTokens.AT
+import org.jetbrains.kotlin.lexer.KtTokens.CLASS_KEYWORD
+import org.jetbrains.kotlin.lexer.KtTokens.COLON
+import org.jetbrains.kotlin.lexer.KtTokens.COLONCOLON
+import org.jetbrains.kotlin.lexer.KtTokens.COMMA
+import org.jetbrains.kotlin.lexer.KtTokens.DOT
+import org.jetbrains.kotlin.lexer.KtTokens.EQ
+import org.jetbrains.kotlin.lexer.KtTokens.EQEQ
+import org.jetbrains.kotlin.lexer.KtTokens.EQEQEQ
+import org.jetbrains.kotlin.lexer.KtTokens.EXCLEQ
+import org.jetbrains.kotlin.lexer.KtTokens.EXCLEQEQEQ
+import org.jetbrains.kotlin.lexer.KtTokens.GT
+import org.jetbrains.kotlin.lexer.KtTokens.GTEQ
+import org.jetbrains.kotlin.lexer.KtTokens.LBRACKET
+import org.jetbrains.kotlin.lexer.KtTokens.LT
+import org.jetbrains.kotlin.lexer.KtTokens.LTEQ
+import org.jetbrains.kotlin.lexer.KtTokens.MUL
+import org.jetbrains.kotlin.lexer.KtTokens.NULL_KEYWORD
+import org.jetbrains.kotlin.lexer.KtTokens.OROR
+import org.jetbrains.kotlin.lexer.KtTokens.QUEST
+import org.jetbrains.kotlin.lexer.KtTokens.RBRACKET
+import org.jetbrains.kotlin.lexer.KtTokens.SAFE_ACCESS
+import org.jetbrains.kotlin.lexer.KtTokens.SEMICOLON
+
+
+internal
+val kotlinGrammar: KotlinGrammar by lazy(LazyThreadSafetyMode.PUBLICATION) {
+    KotlinGrammar()
+}
+
+
+@Suppress("MemberVisibilityCanBePrivate")
+internal
+class KotlinGrammar : Combinator(true, true) {
+
+    // basic building blocks
+    val simpleIdentifier by debug {
+        symbol()
+    }
+
+    val label by debug {
+        simpleIdentifier * token(AT)
+    }
+
+    val comparisonOperator by debug {
+        token(LT) + token(GT) + token(LTEQ) + token(GTEQ)
+    }
+
+    val equalityOperator by debug {
+        token(EXCLEQ) + token(EXCLEQEQEQ) + token(EQEQ) + token(EQEQEQ)
+    }
+
+    val memberAccessOperator by debug {
+        token(DOT) + token(SAFE_ACCESS) + token(COLONCOLON)
+    }
+
+    val semi =
+        token(SEMICOLON)
+
+    val semis =
+        oneOrMore(semi)
+
+
+    val literalConstant =
+        booleanLiteral + integerLiteral + floatLiteral + characterLiteral + token(NULL_KEYWORD)
+
+
+    // recursive parsers
+    var type by reference<Any>()
+
+    var annotation by reference<Any>()
+
+    var fileAnnotation by reference<Any>()
+
+    var parenthesizedUserType by reference<Any>()
+
+    var expression by reference<Any>()
+
+    var primaryExpression by reference<Any?>()
+
+    var directlyAssignableExpression by reference<Any>()
+
+    var assignableExpression by reference<Any>()
+
+
+    // regular parsers
+    val collectionLiteral =
+        bracket(
+            optional(
+                expression * zeroOrMore(token(COMMA) * expression) * optional(token(COMMA))
```

Comment:
🚀 Reuse `token(COMMA)`. In general, token parsers should be reused.

Reply:
Added memoization for simple token parsers.